src/vs/workbench/services/workspaces/common/workspaceTrust.ts

```diff
@@ -386,31 +416,41 @@ export class WorkspaceTrustService extends Disposable implements IWorkspaceTrust
 	}
 
 	async requireWorkspaceTrust(request: WorkspaceTrustRequest = { modal: true }): Promise<WorkspaceTrustState> {
+		// Trusted workspace
 		if (this.currentTrustState === WorkspaceTrustState.Trusted) {
 			return this.currentTrustState;
 		}
+		// Untrusted workspace - soft request
 		if (this.currentTrustState === WorkspaceTrustState.Untrusted && !request.modal) {
 			return this.currentTrustState;
 		}
 
-		if (this._trustRequestPromise) {
-			if (request.modal &&
-				this.requestModel.trustRequest &&
-				!this.requestModel.trustRequest.modal) {
-				this.requestModel.initiateRequest(request);
+		if (request.modal) {
+			// Modal request
+			if (!this._modalTrustRequestPromise) {
+				// Create promise
+				this._modalTrustRequestPromise = new Promise((resolve, reject) => {
+					this._modalTrustRequestResolver = resolve;
+					this._modalTrustRequestRejecter = reject;
+				});
+			} else {
+				// Return existing promises
+				return this._modalTrustRequestPromise;
+			}
+		} else {
+			// Soft soft
+			if (!this._trustRequestPromise) {
+				this._trustRequestPromise = new Promise(resolve => {
+					this._inFlightResolver = resolve;
+				});
+				this._ctxWorkspaceTrustPendingRequest.set(true);
```

Comment:
Is the behavior now that we only show the soft indication and status bar entry if a soft request is made? Personally I feel if the user is asked via modal and they choose a "let me think about it" option, we should keep a pending soft notification out, this way the user can get back to the trust manager and are aware their cancellation prevented some behavior.

Reply:
This one felt a bit weird to me as well. I moved this line so that it applies to both notification types.